### PR TITLE
Add certificate operations

### PR DESCRIPTION
**Pull request by bernard-wagner**
Implements #20 

```
go test -run TestCertificate
PASS
```

### CODE REVIEW

File: certificate.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+//
+// Either or both label and/or id can be nil, in which case random values will be generated.
+func ImportCertificateOnSession(session *PKCS11Session, slot uint, id []byte, label []byte, certificate *x509.Certificate) error {
+	attributes := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_CERTIFICATE),
+		pkcs11.NewAttribute(pkcs11.CKA_CERTIFICATE_TYPE, pkcs11.CKC_X_509),
+		pkcs11.NewAttribute(pkcs11.CKA_TOKEN, true),
+		pkcs11.NewAttribute(pkcs11.CKA_PRIVATE, false),
+		pkcs11.NewAttribute(pkcs11.CKA_SUBJECT, certificate.RawSubject),
+		pkcs11.NewAttribute(pkcs11.CKA_ISSUER, []byte(certificate.RawIssuer)),
```
Comment (nickrmc83) [inline certificate.go]: `RawIssuer` is already a slice of bytes. Is the `[]byte(...)` necessary?

File: certificate.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	err = withSession(slot, func(session *PKCS11Session) error {
+		return ImportCertificateOnSession(session, slot, id, label, certificate)
+	})
+	return err
+}
+
+// ImportCertificateOnSession imports a certificate and populates the available attributes
+//
+// Either or both label and/or id can be nil, in which case random values will be generated.
```
Comment (nickrmc83) [inline certificate.go]: > Either or both label and/or id can be nil, in which case random values will be generated.

I cannot see anywhere in the PKCS#11 specification or in the code below where either the label or id have random values generated when they are nil. Is this a PKCS#11 implementation specific observation or have I missed something in the spec? I've been using [v2.20](https://www.cryptsoft.com/pkcs11doc/STANDARD/pkcs-11v2-20.pdf).

If this is an implementation specific behaviour it'd be nice if we normalised and generated values in crypto11 to be consistent across providers.

**Review by nickrmc83 — CHANGES_REQUESTED** [review]
Generally looks good. @bernard-wagner if you could respond particularly in regards to the behaviour of ID and label setting that'd be appreciated.

File: certificate.go (hunk 2 above)
Comment (nickrmc83) [reply certificate.go]: I think you may want to call `generateKeyLabel()` when a label or id is not set.

Comment (bernard-wagner) [reply certificate.go]: I made the change to generate the label and id if it is not set.

**Comment by dmjones** [comment]
Per my comments in #20, we would be interested in merging in half of these changes. Would you be interested in splitting up your work so that we can merge in the "certificate finding" code, but not the import code?

**Comment by bernard-wagner** [comment]
> Per my comments in #20, we would be interested in merging in half of these changes. Would you be interested in splitting up your work so that we can merge in the "certificate finding" code, but not the import code?

Sorry for the delay. How would you suggest I structure the unit tests without the ability to create or import a certificate?

**Comment by dmjones** [comment]
> Sorry for the delay. How would you suggest I structure the unit tests without the ability to create or import a certificate?

Your original import/creation code could become test helper functions. I would move them to `certificate_test.go` and use them to create your test fixtures.

Over time, more of our tests will probably use those helper functions (as we deprecate and eventually remove key generation).

**Comment by bernard-wagner** [comment]
> > Sorry for the delay. How would you suggest I structure the unit tests without the ability to create or import a certificate?
> 
> Your original import/creation code could become test helper functions. I would move them to `certificate_test.go` and use them to create your test fixtures.
> 
> Over time, more of our tests will probably use those helper functions (as we deprecate and eventually remove key generation).

Made the change as requested.

**Comment by bernard-wagner** [comment]
Rebased the work on v1, which was far simpler. @dmjones possible to restart the review process?

**Comment by dmjones** [comment]
Thanks for updating this. Just to set some expectations, I hope to be able to review this in the first week of August. Apologies for the delay. If I can get to it before then, I will.

**Comment by dmjones** [comment]
Thanks for the PR updates. It looks great, I've made minor adjustments and added a few extra features in #52. I'll close this PR now.

**Pull request closed without merging**